File: ports/python/botiva/state.py
"""State & history ports (PROTOCOL.md §6, §8) — asyncio implementations."""

from __future__ import annotations

from typing import Any, Protocol

from .protocol import Frame


class StateStore(Protocol):
    async def get(self, key: str) -> Any: ...
    async def set(self, key: str, value: Any) -> None: ...
    async def delete(self, key: str) -> None: ...


class MemoryStateStore:
    def __init__(self) -> None:
        self._map: dict[str, Any] = {}

    async def get(self, key: str) -> Any:
        return self._map.get(key)

    async def set(self, key: str, value: Any) -> None:
        self._map[key] = value

    async def delete(self, key: str) -> None:
        self._map.pop(key, None)


class ScopedStore:
    """A namespaced view over one StateStore key."""

    def __init__(self, store: StateStore, key: str) -> None:
        self._store = store
        self.key = key

    async def get(self) -> dict[str, Any] | None:
        return await self._store.get(self.key)

    async def set(self, value: dict[str, Any]) -> None:
        await self._store.set(self.key, value)

    async def patch(self, partial: dict[str, Any]) -> dict[str, Any]:
        """Shallow-merge; ``None`` values delete keys."""
        current = dict(await self.get() or {})
        for k, v in partial.items():
            if v is None:
                current.pop(k, None)
            else:
                current[k] = v
        await self.set(current)
        return current

    async def delete(self) -> None:
        await self._store.delete(self.key)


class UserStore(ScopedStore):
    """Per-user state — survives conversations, devices and reconnects."""

    def __init__(self, store: StateStore, user_id: str) -> None:
        super().__init__(store, f"user:{user_id}")
        self.user_id = user_id


class ConversationStore(ScopedStore):
    """Per-conversation state — shared by every attached connection."""

    def __init__(self, store: StateStore, conversation_id: str) -> None:
        super().__init__(store, f"conv:{conversation_id}")
        self.conversation_id = conversation_id


class HistoryStore(Protocol):
    async def append(self, conversation_id: str, frame: Frame) -> int: ...
    async def after(self, conversation_id: str, watermark: int) -> list[Frame]: ...
    async def latest(self, conversation_id: str) -> int: ...


class MemoryHistoryStore:
    def __init__(self, max_frames: int = 1000) -> None:
        self._max_frames = max_frames
        self._convs: dict[str, dict[str, Any]] = {}

    async def append(self, conversation_id: str, frame: Frame) -> int:
        conv = self._convs.setdefault(conversation_id, {"base_seq": 0, "frames": []})
        seq = conv["base_seq"] + len(conv["frames"]) + 1
        conv["frames"].append({**frame, "seq": seq})
        while len(conv["frames"]) > self._max_frames:
            conv["frames"].pop(0)
            conv["base_seq"] += 1
        return seq

    async def after(self, conversation_id: str, watermark: int) -> list[Frame]:
        conv = self._convs.get(conversation_id)
        if not conv:
            return []
        return [f for f in conv["frames"] if f.get("seq", 0) > watermark]

    async def latest(self, conversation_id: str) -> int:
        conv = self._convs.get(conversation_id)
        return conv["base_seq"] + len(conv["frames"]) if conv else 0

File: ports/dotnet/Botiva/State.cs
using System.Collections.Concurrent;
using System.Text.Json.Nodes;

namespace Botiva;

/// <summary>Persistence port (PROTOCOL.md §8). Values are JSON objects.</summary>
public interface IStateStore
{
    Task<JsonObject?> GetAsync(string key, CancellationToken ct = default);
    Task SetAsync(string key, JsonObject value, CancellationToken ct = default);
    Task DeleteAsync(string key, CancellationToken ct = default);
}

public sealed class MemoryStateStore : IStateStore
{
    private readonly ConcurrentDictionary<string, string> _map = new();

    public Task<JsonObject?> GetAsync(string key, CancellationToken ct = default) =>
        Task.FromResult(_map.TryGetValue(key, out var raw) ? JsonNode.Parse(raw) as JsonObject : null);

    public Task SetAsync(string key, JsonObject value, CancellationToken ct = default)
    {
        _map[key] = value.ToJsonString();
        return Task.CompletedTask;
    }

    public Task DeleteAsync(string key, CancellationToken ct = default)
    {
        _map.TryRemove(key, out _);
        return Task.CompletedTask;
    }
}

/// <summary>A namespaced JSON-object view over one StateStore key.</summary>
public class ScopedStore(IStateStore store, string key)
{
    public string Key { get; } = key;

    public Task<JsonObject?> GetAsync(CancellationToken ct = default) => store.GetAsync(Key, ct);

    public Task SetAsync(JsonObject value, CancellationToken ct = default) => store.SetAsync(Key, value, ct);

    /// <summary>Shallow-merge into the current value; null values delete keys.</summary>
    public async Task<JsonObject> PatchAsync(JsonObject partial, CancellationToken ct = default)
    {
        var current = await GetAsync(ct) ?? new JsonObject();
        foreach (var (k, v) in partial)
        {
            if (v is null) current.Remove(k);
            else current[k] = v.DeepClone();
        }
        await SetAsync(current, ct);
        return current;
    }

    public Task DeleteAsync(CancellationToken ct = default) => store.DeleteAsync(Key, ct);
}

/// <summary>Per-user state — survives conversations, devices and reconnects.</summary>
public sealed class UserStore(IStateStore store, string userId) : ScopedStore(store, $"user:{userId}")
{
    public string UserId { get; } = userId;
}

/// <summary>Per-conversation state — shared by every attached connection.</summary>
public sealed class ConversationStore(IStateStore store, string conversationId)
    : ScopedStore(store, $"conv:{conversationId}")
{
    public string ConversationId { get; } = conversationId;
}

/// <summary>Transcript port (PROTOCOL.md §8).</summary>
public interface IHistoryStore
{
    Task<int> AppendAsync(string conversationId, Frame frame, CancellationToken ct = default);
    Task<IReadOnlyList<Frame>> AfterAsync(string conversationId, int watermark, CancellationToken ct = default);
    Task<int> LatestAsync(string conversationId, CancellationToken ct = default);
}

public sealed class MemoryHistoryStore(int maxFrames = 1000) : IHistoryStore
{
    private sealed class Entry
    {
        public int BaseSeq;
        public readonly List<string> Frames = [];
    }

    private readonly ConcurrentDictionary<string, Entry> _convs = new();

    public Task<int> AppendAsync(string conversationId, Frame frame, CancellationToken ct = default)
    {
        var entry = _convs.GetOrAdd(conversationId, _ => new Entry());
        lock (entry)
        {
            var seq = entry.BaseSeq + entry.Frames.Count + 1;
            var stored = (Frame)frame.DeepClone();
            stored["seq"] = seq;
            entry.Frames.Add(stored.ToJsonString());
            while (entry.Frames.Count > maxFrames)
            {
                entry.Frames.RemoveAt(0);
                entry.BaseSeq++;
            }
            return Task.FromResult(seq);
        }
    }

    public Task<IReadOnlyList<Frame>> AfterAsync(string conversationId, int watermark, CancellationToken ct = default)
    {
        if (!_convs.TryGetValue(conversationId, out var entry))
            return Task.FromResult<IReadOnlyList<Frame>>([]);
        lock (entry)
        {
            var result = entry.Frames
                .Select(raw => (Frame)JsonNode.Parse(raw)!)
                .Where(f => (int?)f["seq"] > watermark)
                .ToList();
            return Task.FromResult<IReadOnlyList<Frame>>(result);
        }
    }

    public Task<int> LatestAsync(string conversationId, CancellationToken ct = default)
    {
        if (!_convs.TryGetValue(conversationId, out var entry)) return Task.FromResult(0);
        lock (entry)
        {
            return Task.FromResult(entry.BaseSeq + entry.Frames.Count);
        }
    }
}
